src/self_driving/planning/pilot.cpp

```diff
@@ -32,27 +50,310 @@ Pilot::Pilot(std::string model_save_path, std::string forecast_model_save_path,
+      // have already been captured by out_metadata and out_params.
+      //
+      // Under the assumption that the task manager is not backlogged by tasks
+      // submitted by QueryTraceMetricRawData to write frequency information,
+      // the frequency information we pull from the tables should also be
+      // reasonably up to date.
+    }
+  }
+
+  // This flag is used to make sure the 2 separate stages (get forecast prediction, construct
+  // WorkloadForecast object) are consistent. "Consistency" here means that all the data comes
+  // from either the disk file or internal tables.
+  bool loaded_from_internal = false;
+  std::pair<selfdriving::WorkloadForecastPrediction, bool> result;
+  {
+    // Only get the most recent interval information
+    bool success = false;
+    std::vector<std::string> models{"LSTM"};
+    std::unordered_map<int64_t, std::vector<double>> segment_information;
+    if (metrics_in_db && task_manager_) {
+      segment_information = GetSegmentInformation(ComputeTimestampDataRange(timestamp), &success);
+    }
+
+    if (segment_information.empty() || !success) {
+      loaded_from_internal = false;
+      std::string input_path{metrics::QueryTraceMetricRawData::FILES[1]};
+
+      // If the segment information is empty, use the file instead on disk
+      result =
+          model_server_manager_->InferForecastModel(input_path, forecast_model_save_path_, models, nullptr,
+                                                    workload_forecast_interval_, sequence_length_, horizon_length_);
+    } else {
+      loaded_from_internal = true;
+      result =
+          model_server_manager_->InferForecastModel(&segment_information, forecast_model_save_path_, models, nullptr,
+                                                    workload_forecast_interval_, sequence_length_, horizon_length_);
+    }
+
+    if (!result.second) {
+      SELFDRIVING_LOG_ERROR("Forecast model inference failed");
+      metrics_thread_->ResumeMetrics();
+      return;
+    }
+  }
+
+  if (task_manager_ && metrics_in_db && loaded_from_internal) {
+    // Retrieve query information from internal tables
+    auto metadata_result = RetrieveWorkloadMetadata(ComputeTimestampDataRange(timestamp), out_metadata, out_params);
+    if (!metadata_result.second) {
+      SELFDRIVING_LOG_ERROR("Failed to read from internal trace metadata tables");
+      metrics_thread_->ResumeMetrics();
+      return;
+    }
+
+    // Record forecast into internal tables
+    RecordWorkloadForecastPrediction(timestamp, result.first, metadata_result.first);
+
+    // Construct workload forecast
+    forecast_ = std::make_unique<selfdriving::WorkloadForecast>(result.first, std::move(metadata_result.first));
+  } else {
+    auto sample = settings_manager_->GetInt(settings::Param::forecast_sample_limit);
+    forecast_ = std::make_unique<selfdriving::WorkloadForecast>(workload_forecast_interval_, sample);
```

Comment:
There should probably be three different options for `LoadWorkloadForecast`: (1) read from internal tables to do inference, (2) read from the disk to do inference, and (3) directly read the query trace from the disk without inference. All three scenarios could be useful for us. And right now if we're not in case (1), we'll do both (2) and (3) together and just use the results from (3). This could be confusing.

Reply:
Added 3 separate flags that essentially does:
- `INTERNAL_TABLES_WITH_INFERENCE`: pulls all data from internal tables and records inference results
- `DISK_WITH_INFERENCE`: inference from disk file and construct workload forecast from inference results (no recording to internal tables)
- `DISK_ONLY`: construct workload forecast without any inference